Webpack DLLPlugin

I don't know if this is really an issue, more of a question.

I am trying to utilize the [Webpack DLLPlugin](https://webpack.js.org/plugins/dll-plugin/) to build a manifest of my API. I don't know if it's a workflow thing or something else. The DLLPlugin samples make it look straight forward/

Here is where I create the DLLs using the sample app provided.

```js
// dll.webpack.config.js
const DojoWebpackPlugin = require("dojo-webpack-plugin");
const path = require("path");
const webpack = require("webpack");

module.exports = {
  resolve: {
    extensions: [".js", ".css"]
  },
  entry: {
    commons: ["dojo/_base/lang", "dojo/_base/declare"],
    dijit: ["dijit/layout/TabContainer", "dijit/layout/ContentPane"]
  },
  output: {
    path: path.join(__dirname, "dist"),
    filename: "dojo.[name].js",
    library: "[name]_[hash]"
  },
  plugins: [
    new webpack.DllPlugin({
      path: path.join(__dirname, "dist", "[name]-manifest.json"),
      name: "[name]_[hash]"
    }),
		new DojoWebpackPlugin({
			loaderConfig: require("./js/loaderConfig"),
			environment: {dojoRoot: "dist"},	// used at run time for non-packed resources (e.g. blank.gif)
			buildEnvironment: {dojoRoot: "node_modules"}, // used at build time
			locales: ["en"]
		})
  ],
};
```

My DLLs get generated fine. I can check the output and see my dojo/dijit deps just fine.

Then I try to use them here.

```js
// webpack.config.js
...
	plugins: [
		new webpack.DllReferencePlugin({
			context: path.join(__dirname, "..", "dist"),
			manifest: require("./dist/commons-manifest.json")
		}),
		new webpack.DllReferencePlugin({
			context: path.join(__dirname, "..", "dist"),
			manifest: require("./dist/dijit-manifest.json")
		}),
		new DojoWebpackPlugin({
			loaderConfig: require("./js/loaderConfig"),
			environment: {dojoRoot: "dist"},	// used at run time for non-packed resources (e.g. blank.gif)
			buildEnvironment: {dojoRoot: "node_modules"}, // used at build time
			locales: ["en"]
		})
        ],
...
```

No errors, but the DLLs don't seem to get utilized. I've tried switching the order of DLL refs and dojo-webpack-plugin around and no luck.

I was expecting that Webpack would see the output the dojo-webpack-plugin is passing into the build and utilize the DLLs, but maybe that's not how it works.

I was curious if there was something I was missing related to the dojo-webpack-plugin and how it's used or if anyone has tried this before.

Thanks!

I thought I would just update where I ended up with creating DLLs and using the plugin.

I got it to work, sorta.

The key was `context` and `globalContext`.

```js
 new webpack.DllPlugin({
  context: path.join(__dirname, "node_modules", "dojo"),
  path: path.join(__dirname, "dist", "[name]-manifest.json"),
  name: "[name]_[hash]"
}),
new DojoWebpackPlugin({
  globalContext: path.join(__dirname, "node_modules", "dojo"),
  loaderConfig: require("./js/loaderConfig"),
  environment: { dojoRoot: "dist" },
  buildEnvironment: { dojoRoot: "node_modules" },
  locales: ["en"]
});

// Then I can use it
new webpack.DllReferencePlugin({
  context: path.join(__dirname, "node_modules", "dojo"),
  manifest: require("./dist/dojo-manifest.json")
}),
new DojoWebpackPlugin({
  loaderConfig: require("./js/loaderConfig"),
  environment: { dojoRoot: "dist" },
  buildEnvironment: { dojoRoot: "node_modules" },
  locales: ["en"]
})
```

That creates the DLLs fine and I can use them and they do get used. But I get `Module not found: dojo/has` once I do.

It was interesting experiment and I'll keep seeing what I can come up with, but progress is progress. If I can get the DLLPlugin working better with the dojo-webpack-plugin, I'll update. Thanks!

I tried building the https://github.com/odoe/dojo-webpack-plugin-sample-dll, but I get the following error:
```
C:\git\dojo-webpack-plugin-sample-dll\node_modules\webpack-cli\bin\webpack.js:242
                                throw err;
                                ^

Error: Cannot find module './dist/dojo-manifest.json'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
```

Did you forget to include some files?

--- a/lib/DojoAMDModuleFactoryPlugin.js
+++ b/lib/DojoAMDModuleFactoryPlugin.js
@@ -19,6 +19,7 @@ const {newNormalModule} = require("./compat");
 const DojoAMDRequireItemDependency = require("./DojoAMDRequireItemDependency");
 const {reg, tap, callSync, callSyncBail} = require("./pluginCompat").for("dojo-webpack-plugin");
 const SingleEntryDependency = require("webpack/lib/dependencies/SingleEntryDependency");
+const Module = require("webpack/lib/Module");
 
 module.exports = class DojoAMDModuleFactoryPlugin {
 	constructor(options) {
@@ -210,13 +211,20 @@ module.exports = class DojoAMDModuleFactoryPlugin {
 	}
 
 	module(module) {
+		if (module.originalRequest instanceof Module) {
+			// Copy absMids from original request to the new request
+			callSync(this.factory, "filter absMids", module.originalRequest, absMid => {
+				callSync(this.factory, "add absMid", module, absMid);
+				return true;
+			});
+		}
 		// If the module already exists in the compilation, then copy the absMid data from
 		// this module to the existing module since this module will be discarded
 		const existing = this.compilation.findModule(module.request);
-		if (existing && module.absMid) {
-			callSync(this.factory, "add absMid", existing, module.absMid);
-			(module.absMidAliases||[]).forEach(absMid => {
+		if (existing) {
+			callSync(this.factory, "filter absMids", module, absMid => {
 				callSync(this.factory, "add absMid", existing, absMid);
+				return true;
 			});
 		}
 		// Add functions to the module object for adding/filtering absMids (for use by loaders)

--- a/test/DojoAMDModuleFactoryPlugin.test.js
+++ b/test/DojoAMDModuleFactoryPlugin.test.js
@@ -8,6 +8,7 @@
  */
 const DojoAMDModuleFactoryPlugin = require("../lib/DojoAMDModuleFactoryPlugin");
 const {Tapable, reg, callSync, callSyncWaterfall} = require("../lib/pluginCompat");
+const Module = require("webpack/lib/Module");
 const plugin = new DojoAMDModuleFactoryPlugin({});
 
 class Factory extends Tapable {
@@ -202,6 +203,19 @@ describe("DojoAMDModuleFactoryPlugin tests", function() {
 			(typeof result.filterAbsMids).should.be.eql('function');
 			existing.absMid.should.eql('a');
 			existing.absMidAliases.length.should.eql(0);
+		});
+
+		it("Should copy absMids from original request of delgated modules", function() {
+			const originalModule = new Module("test");
+			const delegated = {originalRequest: originalModule};
+			callSync(factory, "add absMid", originalModule, "a");
+			compilation.findModule = function() { return null; };
+			debugger; // eslint-disable-line
+			const result = callSyncWaterfall(factory, "module", delegated);
+			result.should.be.eql(delegated);
+			result.absMid.should.eql('a');
+			result.absMidAliases.length.should.eql(0);
+			originalModule.absMid.should.eql('a');
 		});
 	});
 

